[RN 0.71.0 build error] Expected directory '.../node_modules/react-native/android' to contain exactly one file, however, it contains no files

Hey there, thanks a lot for all the work on this 🙏 

I came across the following when I updated a project to 0.71.0 when building android:
```
* Where:
Build file '.../node_modules/react-native-random-values-jsi-helper/android/build.gradle' line: 172

* What went wrong:
A problem occurred evaluating project ':react-native-random-values-jsi-helper'.
> Expected directory '.../node_modules/react-native/android' to contain exactly one file, however, it contains no files.
```

seems to be related to https://github.com/react-native-community/discussions-and-proposals/pull/508

also, they added this to react-native/android/README.md

![Screenshot 2023-01-19 at 05 49 33](https://user-images.githubusercontent.com/11508521/213344048-c57a6892-11e5-4258-826c-0ebfb095b12c.png)

I tried to fix it, but then came across another issue related to CMakeLists.txt, 
`[CMake Error: The following variables are used in this project, but they are set to NOTFOUND]` for REACT_NATIVE_JNI_LIB and JSI_LIB

probably was trying the wrong thing. 😄

--- a/android/CMakeLists.txt
+++ b/android/CMakeLists.txt
@@ -1,64 +1,36 @@
+project(reactnativerandomvaluesjsihelper)
 cmake_minimum_required(VERSION 3.9.0)
 
 set (PACKAGE_NAME "react-native-random-values-jsi-helper")
 set (BUILD_DIR ${CMAKE_SOURCE_DIR}/build)
+set(CMAKE_VERBOSE_MAKEFILE ON)
+set(CMAKE_CXX_STANDARD 17)
 
 include_directories(
         "../cpp"
-        "${NODE_MODULES_DIR}/react-native/React"
-        "${NODE_MODULES_DIR}/react-native/React/Base"
-        "${NODE_MODULES_DIR}/react-native/ReactCommon/jsi"
 )
-
-if(${REACT_NATIVE_VERSION} LESS 66)
-  set (
-          INCLUDE_JSI_CPP
-          "${NODE_MODULES_DIR}/react-native/ReactCommon/jsi/jsi/jsi.cpp"
-  )
-endif()
 
 add_library(reactnativerandomvaluesjsihelper
         SHARED
         cpp-adapter.cpp
         "../cpp/TypedArray.cpp"
-        ${INCLUDE_JSI_CPP}
 )
 set_target_properties(
         reactnativerandomvaluesjsihelper PROPERTIES
-        CXX_STANDARD 14
+        CXX_STANDARD 17
         CXX_EXTENSIONS OFF
         POSITION_INDEPENDENT_CODE ON
 )
 
-file (GLOB LIBRN_DIR "${PREBUILT_DIR}/${ANDROID_ABI}")
-
+find_package(ReactAndroid REQUIRED CONFIG)
 find_library(
         log-lib
         log
 )
-find_library(
-        REACT_NATIVE_JNI_LIB
-        reactnativejni
-        PATHS ${LIBRN_DIR}
-        NO_CMAKE_FIND_ROOT_PATH
-)
-if(${REACT_NATIVE_VERSION} LESS 66)
-  # JSI lib didn't exist on RN 0.65 and before. Simply omit it.
-  set (JSI_LIB "")
-else()
-  # RN 0.66 distributes libjsi.so, can be used instead of compiling jsi.cpp manually.
-  find_library(
-          JSI_LIB
-          jsi
-          PATHS ${LIBRN_DIR}
-          NO_CMAKE_FIND_ROOT_PATH
-  )
-endif()
 
 target_link_libraries(
         reactnativerandomvaluesjsihelper
         ${log-lib}
-        ${JSI_LIB}
-        ${REACT_NATIVE_JNI_LIB}
+        ReactAndroid::jsi 
         android
 )

--- a/android/build.gradle
+++ b/android/build.gradle
@@ -1,47 +1,29 @@
 import org.apache.tools.ant.filters.ReplaceTokens
 import java.nio.file.Paths
 
-static def findNodeModules(baseDir) {
-  def basePath = baseDir.toPath().normalize()
-  // Node's module resolution algorithm searches up to the root directory,
-  // after which the base path will be null
-  while (basePath) {
-    def nodeModulesPath = Paths.get(basePath.toString(), "node_modules")
-    def reactNativePath = Paths.get(nodeModulesPath.toString(), "react-native")
-    if (nodeModulesPath.toFile().exists() && reactNativePath.toFile().exists()) {
-      return nodeModulesPath.toString()
-    }
-    basePath = basePath.getParent()
-  }
-  throw new GradleException("BlobJsiHelper: Failed to find node_modules/ path!")
-}
-
-def nodeModules = findNodeModules(projectDir)
-logger.warn("BlobJsiHelper: node_modules/ found at: ${nodeModules}")
-
 buildscript {
     repositories {
         google()
         mavenCentral()
-        jcenter()
     }
-
 
     dependencies {
-        classpath 'com.android.tools.build:gradle:4.2.2'
-        classpath 'de.undercouch:gradle-download-task:4.1.2'
+        classpath 'com.android.tools.build:gradle:7.2.2'
     }
 }
 
-apply plugin: 'com.android.library'
-apply plugin: 'de.undercouch.download'
-
-def getExtOrDefault(name) {
-  return rootProject.ext.has(name) ? rootProject.ext.get(name) : project.properties['RandomValuesJsiHelper_' + name]
+def resolveBuildType() {
+    Gradle gradle = getGradle()
+    String tskReqStr = gradle.getStartParameter().getTaskRequests()['args'].toString()
+    return tskReqStr.contains('Release') ? 'release' : 'debug'
 }
 
-def getExtOrIntegerDefault(name) {
-  return rootProject.ext.has(prop) ? rootProject.ext.get(prop) : fallback
+def isNewArchitectureEnabled() {
+  // To opt-in for the New Architecture, you can either:
+  // - Set `newArchEnabled` to true inside the `gradle.properties` file
+  // - Invoke gradle with `-newArchEnabled=true`
+  // - Set an environment variable `ORG_GRADLE_PROJECT_newArchEnabled=true`
+  return project.hasProperty("newArchEnabled") && project.newArchEnabled == "true"
 }
 
 def reactNativeArchitectures() {
@@ -49,43 +31,23 @@ def reactNativeArchitectures() {
   return value ? value.split(",") : ["armeabi-v7a", "x86", "x86_64", "arm64-v8a"]
 }
 
-def sourceBuild = false
-def defaultDir = null
-def androidSourcesDir = null
-def androidSourcesName = 'React Native sources'
-
-if (rootProject.ext.has('reactNativeAndroidRoot')) {
-  defaultDir = rootProject.ext.get('reactNativeAndroidRoot')
-  androidSourcesDir = defaultDir.parentFile.toString()
-} else if (findProject(':ReactAndroid') != null) {
-  sourceBuild = true
-  defaultDir = project(':ReactAndroid').projectDir
-  androidSourcesDir = defaultDir.parentFile.toString()
-} else {
-  defaultDir = file("$nodeModules/react-native/android")
-  androidSourcesDir = defaultDir.parentFile.toString()
-}
-
-if (!defaultDir.exists()) {
-  throw new GradleException(
-    "${project.name}: React Native android directory (node_modules/react-native/android) does not exist! Resolved node_modules to: ${nodeModules}"
-  )
-}
-def prebuiltDir = sourceBuild
-  ? "$nodeModules/react-native/ReactAndroid/src/main/jni/prebuilt/lib"
-  : "$buildDir/react-native-0*/jni"
-
 def safeExtGet(prop, fallback) {
   rootProject.ext.has(prop) ? rootProject.ext.get(prop) : fallback
 }
 
+if (isNewArchitectureEnabled()) {
+  apply plugin: 'com.facebook.react'
+}
+apply plugin: 'com.android.library'
 
-def reactProperties = new Properties()
-file("$nodeModules/react-native/ReactAndroid/gradle.properties").withInputStream { reactProperties.load(it) }
-def REACT_NATIVE_VERSION = reactProperties.getProperty("VERSION_NAME").split("\\.")[1].toInteger()
 android {
   compileSdkVersion safeExtGet('compileSdkVersion', 30)
   ndkVersion '21.4.7075529'
+
+  buildFeatures {
+    prefab true
+  }
+
   defaultConfig {
         minSdkVersion safeExtGet('minSdkVersion', 16)
         targetSdkVersion safeExtGet('targetSdkVersion', 30)
@@ -94,21 +56,20 @@ android {
 
         externalNativeBuild {
           cmake {
-            cppFlags "-fexceptions", "-frtti", "-std=c++1y", "-DONANDROID"
-            arguments '-DANDROID_STL=c++_shared',
-              "-DREACT_NATIVE_VERSION=${REACT_NATIVE_VERSION}",
-              "-DNODE_MODULES_DIR=${nodeModules}",
-              "-DPREBUILT_DIR=${prebuiltDir}"
+            cppFlags "-O2 -frtti -fexceptions -Wall -Wno-unused-variable -fstack-protector-all"
+            arguments '-DANDROID_STL=c++_shared'
+            abiFilters (*reactNativeArchitectures())
           }
         }
-        ndk {
-          abiFilters (*reactNativeArchitectures())
+
+        packagingOptions {
+            doNotStrip resolveBuildType() == 'debug' ? "**/**/*.so" : ''
+            excludes = [
+                    "META-INF",
+                    "META-INF/**",
+                    "**/libjsi.so",
+            ]
         }
-
-    }
-
-    dexOptions {
-      javaMaxHeapSize "4g"
     }
 
     externalNativeBuild {
@@ -117,246 +78,17 @@ android {
         }
     }
 
-    packagingOptions {
-      excludes = ["**/libc++_shared.so", "**/libfbjni.so", "**/libreactnativejni.so", "**/libjsi.so", "**/MANIFEST.MF"]
-    }
-
-    buildTypes {
-        release {
-            minifyEnabled false
-        }
-    }
-    lintOptions {
-        disable 'GradleCompatible'
-    }
     compileOptions {
         sourceCompatibility JavaVersion.VERSION_1_8
         targetCompatibility JavaVersion.VERSION_1_8
     }
-    configurations {
-      extractHeaders
-      extractJNI
-    }
-
 }
 
 repositories {
-    mavenLocal()
-    maven {
-        // All of React Native (JS, Obj-C sources, Android binaries) is installed from npm
-        url("$rootDir/../node_modules/react-native/android")
-    }
-    google()
     mavenCentral()
-    jcenter()
 }
 
 dependencies {
     //noinspection GradleDynamicVersion
-    implementation "com.facebook.react:react-native:+"  // From node_modules
-    extractHeaders("com.facebook.fbjni:fbjni:+:headers")
-    //noinspection GradleDynamicVersion
-    extractJNI("com.facebook.fbjni:fbjni:+")
-
-    if (!sourceBuild) {
-        def buildType = "debug"
-        tasks.all({ task ->
-            if (task.name == "buildCMakeRelease") {
-                buildType = "release"
-            }
-        })
-        def rnAarMatcher = "**/react-native/**/*${buildType}.aar"
-        if (REACT_NATIVE_VERSION < 69) {
-            rnAarMatcher = "**/**/*.aar"
-        }
-        def rnAAR = fileTree("${defaultDir.toString()}").matching({ it.include rnAarMatcher }).singleFile
-        extractJNI(files(rnAAR))
-    }
-}
-
-
-def downloadsDir = new File("$buildDir/downloads")
-def thirdPartyNdkDir = new File("$buildDir/third-party-ndk")
-def thirdPartyVersionsFile = new File("${androidSourcesDir.toString()}/ReactAndroid/gradle.properties")
-def thirdPartyVersions = new Properties()
-thirdPartyVersions.load(new FileInputStream(thirdPartyVersionsFile))
-
-def BOOST_VERSION = thirdPartyVersions["BOOST_VERSION"]
-def boost_file = new File(downloadsDir, "boost_${BOOST_VERSION}.tar.gz")
-def DOUBLE_CONVERSION_VERSION = thirdPartyVersions["DOUBLE_CONVERSION_VERSION"]
-def double_conversion_file = new File(downloadsDir, "double-conversion-${DOUBLE_CONVERSION_VERSION}.tar.gz")
-def FOLLY_VERSION = thirdPartyVersions["FOLLY_VERSION"]
-def folly_file = new File(downloadsDir, "folly-${FOLLY_VERSION}.tar.gz")
-def GLOG_VERSION = thirdPartyVersions["GLOG_VERSION"]
-def glog_file = new File(downloadsDir, "glog-${GLOG_VERSION}.tar.gz")
-
-task createNativeDepsDirectories {
-  doLast {
-    downloadsDir.mkdirs()
-    thirdPartyNdkDir.mkdirs()
-  }
-}
-
-task downloadBoost(dependsOn: createNativeDepsDirectories, type: Download) {
-  def transformedVersion = BOOST_VERSION.replace("_", ".")
-  def srcUrl = "https://boostorg.jfrog.io/artifactory/main/release/${transformedVersion}/source/boost_${BOOST_VERSION}.tar.gz"
-  if (REACT_NATIVE_VERSION < 69) {
-      srcUrl = "https://github.com/react-native-community/boost-for-react-native/releases/download/v${transformedVersion}-0/boost_${BOOST_VERSION}.tar.gz"
-  }
-  src(srcUrl)
-  onlyIfNewer(true)
-  overwrite(false)
-  dest(boost_file)
-}
-
-task prepareBoost(dependsOn: downloadBoost, type: Copy) {
-  from(tarTree(resources.gzip(downloadBoost.dest)))
-  from("src/main/jni/third-party/boost/Android.mk")
-  include("Android.mk", "boost_${BOOST_VERSION}/boost/**/*.hpp", "boost/boost/**/*.hpp")
-  includeEmptyDirs = false
-  into("$thirdPartyNdkDir") // /boost_X_XX_X
-  doLast {
-    file("$thirdPartyNdkDir/boost_${BOOST_VERSION}").renameTo("$thirdPartyNdkDir/boost")
-  }
-}
-
-task downloadDoubleConversion(dependsOn: createNativeDepsDirectories, type: Download) {
-  src("https://github.com/google/double-conversion/archive/v${DOUBLE_CONVERSION_VERSION}.tar.gz")
-  onlyIfNewer(true)
-  overwrite(false)
-  dest(double_conversion_file)
-}
-
-task prepareDoubleConversion(dependsOn: downloadDoubleConversion, type: Copy) {
-  from(tarTree(downloadDoubleConversion.dest))
-  from("src/main/jni/third-party/double-conversion/Android.mk")
-  include("double-conversion-${DOUBLE_CONVERSION_VERSION}/src/**/*", "Android.mk")
-  filesMatching("*/src/**/*", { fname -> fname.path = "double-conversion/${fname.name}" })
-  includeEmptyDirs = false
-  into("$thirdPartyNdkDir/double-conversion")
-}
-
-task downloadFolly(dependsOn: createNativeDepsDirectories, type: Download) {
-  src("https://github.com/facebook/folly/archive/v${FOLLY_VERSION}.tar.gz")
-  onlyIfNewer(true)
-  overwrite(false)
-  dest(folly_file)
-}
-
-task prepareFolly(dependsOn: downloadFolly, type: Copy) {
-  from(tarTree(downloadFolly.dest))
-  from("src/main/jni/third-party/folly/Android.mk")
-  include("folly-${FOLLY_VERSION}/folly/**/*", "Android.mk")
-  eachFile { fname -> fname.path = (fname.path - "folly-${FOLLY_VERSION}/") }
-  includeEmptyDirs = false
-  into("$thirdPartyNdkDir/folly")
-}
-
-task downloadGlog(dependsOn: createNativeDepsDirectories, type: Download) {
-  src("https://github.com/google/glog/archive/v${GLOG_VERSION}.tar.gz")
-  onlyIfNewer(true)
-  overwrite(false)
-  dest(glog_file)
-}
-
-task prepareGlog(dependsOn: downloadGlog, type: Copy) {
-  from(tarTree(downloadGlog.dest))
-  from("src/main/jni/third-party/glog/")
-  include("glog-${GLOG_VERSION}/src/**/*", "Android.mk", "config.h")
-  includeEmptyDirs = false
-  filesMatching("**/*.h.in") {
-    filter(ReplaceTokens, tokens: [
-      ac_cv_have_unistd_h           : "1",
-      ac_cv_have_stdint_h           : "1",
-      ac_cv_have_systypes_h         : "1",
-      ac_cv_have_inttypes_h         : "1",
-      ac_cv_have_libgflags          : "0",
-      ac_google_start_namespace     : "namespace google {",
-      ac_cv_have_uint16_t           : "1",
-      ac_cv_have_u_int16_t          : "1",
-      ac_cv_have___uint16           : "0",
-      ac_google_end_namespace       : "}",
-      ac_cv_have___builtin_expect   : "1",
-      ac_google_namespace           : "google",
-      ac_cv___attribute___noinline  : "__attribute__ ((noinline))",
-      ac_cv___attribute___noreturn  : "__attribute__ ((noreturn))",
-      ac_cv___attribute___printf_4_5: "__attribute__((__format__ (__printf__, 4, 5)))"
-    ])
-    it.path = (it.name - ".in")
-  }
-  into("$thirdPartyNdkDir/glog")
-
-  doLast {
-    copy {
-      from(fileTree(dir: "$thirdPartyNdkDir/glog", includes: ["stl_logging.h", "logging.h", "raw_logging.h", "vlog_is_on.h", "**/src/glog/log_severity.h"]).files)
-      includeEmptyDirs = false
-      into("$thirdPartyNdkDir/glog/exported/glog")
-    }
-  }
-}
-
-task prepareThirdPartyNdkHeaders {
-  if (!boost_file.exists()) {
-    dependsOn(prepareBoost)
-  }
-  if (!double_conversion_file.exists()) {
-    dependsOn(prepareDoubleConversion)
-  }
-  if (!folly_file.exists()) {
-    dependsOn(prepareFolly)
-  }
-  if (!glog_file.exists()) {
-    dependsOn(prepareGlog)
-  }
-}
-
-prepareThirdPartyNdkHeaders.mustRunAfter createNativeDepsDirectories
-
-task extractAARHeaders {
-  doLast {
-    configurations.extractHeaders.files.each {
-      def file = it.absoluteFile
-      copy {
-        from zipTree(file)
-        into "$buildDir/$file.name"
-        include "**/*.h"
-      }
-    }
-  }
-}
-extractAARHeaders.mustRunAfter prepareThirdPartyNdkHeaders
-
-task extractJNIFiles {
-  doLast {
-    configurations.extractJNI.files.each {
-      def file = it.absoluteFile
-
-      copy {
-        from zipTree(file)
-        into "$buildDir/$file.name"
-        include "jni/**/*"
-      }
-    }
-  }
-}
-extractJNIFiles.mustRunAfter extractAARHeaders
-
-def nativeBuildDependsOn(dependsOnTask, variant) {
-  def buildTasks = tasks.findAll({ task ->
-    !task.name.contains("Clean") && (task.name.contains("externalNative") || task.name.contains("CMake")) })
-  if (variant != null) {
-    buildTasks = buildTasks.findAll({ task -> task.name.contains(variant) })
-  }
-  buildTasks.forEach { task -> task.dependsOn(dependsOnTask) }
-}
-
-afterEvaluate {
-  if (sourceBuild) {
-    nativeBuildDependsOn(":ReactAndroid:packageReactNdkDebugLibsForBuck", "Debug")
-    nativeBuildDependsOn(":ReactAndroid:packageReactNdkReleaseLibsForBuck", "Rel")
-  } else {
-    nativeBuildDependsOn(extractAARHeaders, null)
-    nativeBuildDependsOn(extractJNIFiles, null)
-    nativeBuildDependsOn(prepareThirdPartyNdkHeaders, null)
-  }
+    implementation "com.facebook.react:react-android:+"
 }

